src/ToricVarieties/ToricMorphisms/attributes.jl

```diff
@@ -0,0 +1,199 @@
+
+For a given toric morphism `tm`, this method computes the corresponding
+map of the torusinvariant Weil divisors.
+
+# Examples
+```jldoctest
+julia> F4 = hirzebruch_surface(4)
+A normal, non-affine, smooth, projective, gorenstein, non-fano, 2-dimensional toric variety without torusfactor
+
+julia> morphism_on_torusinvariant_weil_divisor_group(ToricIdentityMorphism(F4))
+Map with following data
+Domain:
+=======
+Abelian group with structure: Z^4
+Codomain:
+=========
+Abelian group with structure: Z^4
+```
+"""
+@attr GrpAbFinGenMap function morphism_on_torusinvariant_weil_divisor_group(tm::ToricMorphism)
+
+    # find domain and codomain variety
+    domain_variety = domain(tm)
+    codomain_variety = codomain(tm)
+
+    # find rays of these varieties
+    domain_variety_rays = matrix(ZZ,rays(domain_variety))
+    codomain_variety_rays = matrix(ZZ,rays(codomain_variety))
+
+    # identify the image rays
+    images_of_rays = domain_variety_rays * matrix(grid_morphism(tm))
+
+    # for each image ray, find a max cone in the codomain toric variety which contains it
+    codomain_variety_max_cones = maximal_cones(codomain_variety)
+    rayimage_maxcones_incidence_list = [0 for i in 1:nrows(images_of_rays)]
+    for i in 1:nrows(images_of_rays)
+      for j in 1:length(codomain_variety_max_cones)
+        if contains(codomain_variety_max_cones[j], [images_of_rays[i,k] for k in 1:ncols(images_of_rays)])
+          rayimage_maxcones_incidence_list[i] = j
+          break
+        end
+      end
+    end
+
+    # form list of rays which are in maximal cones of the codomain variety
+    codomain_variety_maxcone_ray_incidence = ray_indices(maximal_cones(codomain_variety))
+    r = rank(character_lattice(codomain_variety))
+    codomain_variety_rays_in_cones = []
+    for i in 1:length(codomain_variety_max_cones)
+        matrix = zeros(Int,nrows(codomain_variety_rays),ncols(codomain_variety_rays))
+        for j in 1:nrows(codomain_variety_rays)
+          if codomain_variety_maxcone_ray_incidence[i,j]
+            for k in 1:ncols(codomain_variety_rays)
+              matrix[j,k] = codomain_variety_rays[j,k]
+            end
+          end
+        end
+        push!(codomain_variety_rays_in_cones, matrix)
+    end
+
+    # compute mapping matrix
+    mapping_matrix = zeros(ZZ, rank(torusinvariant_weil_divisor_group(domain(tm))), rank(torusinvariant_weil_divisor_group(codomain(tm))))
+    for i in 1:nrows(images_of_rays)
+      current_row = images_of_rays[i,:]
+      solution = false
+      matrix_multiplier = 1
+      while !solution
+        r = matrix_multiplier * current_row
+        matrix_multiplier = matrix_multiplier + 1
+        m = matrix(ZZ,codomain_variety_rays_in_cones[rayimage_maxcones_incidence_list[i]])
+        solution = can_solve(transpose(m), transpose(r))
+        if solution
+          current_row = solve(transpose(m), transpose(r))
+          current_row = transpose(current_row)
+        end
+      end
```

Comment:
Why not `while !can_solve(...)`? Also the `matrix_multiplier` seems unneeded. You could just add `images_of_rays[i,:]` to `current_row` and addition might be faster than multiplication.